bubuku/features/rebalance/change_simple.py

```diff
@@ -0,0 +1,443 @@
+
+    def can_accept_partition(self, partition: Partition):
+        for b in partition.brokers:
+            if b.id_ == self.id_:
+                return False
+        return True
+
+
+class ZoneChecker(object):
+    def __init__(self, zones: List[str]):
+        self.zones = zones
+        self.expected_distributions = {}
+
+    def is_allowed_move(self, from_: Broker, to: Broker, partition: Partition):
+        current_zone_distribution = {z: 0 for z in self.zones}
+        for b in partition.brokers:
+            if b.zone not in self.zones:
+                continue  # This actually removes fake one
+            current_zone_distribution[b.zone] = current_zone_distribution.get(b.zone, 0) + 1
+        if from_.zone in current_zone_distribution:
+            current_zone_distribution[from_.zone] -= 1
+        if 0 == current_zone_distribution.get(to.zone, 0):
+            return True
+        # Allow move only to zone with minimal amount of partitions
+        return current_zone_distribution[to.zone] == min(current_zone_distribution.values())
+
+    def _get_expected_distribution(self, count: int) -> List[int]:
+        if count not in self.expected_distributions:
+            min_per_broker = count // len(self.zones)
+            delta = count - min_per_broker * len(self.zones)
+            result = [min_per_broker + 1 if i < delta else min_per_broker for i in range(0, len(self.zones))]
+            self.expected_distributions[count] = result
+        return self.expected_distributions[count].copy()
+
+    def get_improper_brokers(self, brokers) -> List[Broker]:
+        expected = sorted(self._get_expected_distribution(len(brokers)))
+        current_distribution = {}
+        for b in brokers:
+            current_distribution[b.zone] = current_distribution.get(b.zone, 0) + 1
+        broken_zones = {}
+        for zone in [k for k in current_distribution.keys() if k in self.zones]:
+            value = current_distribution[zone]
+            delta = value - expected[-1]
+            if delta >= 0:
+                expected = expected[:-1]
+            if delta > 0:
+                broken_zones[zone] = delta
+        result = []
+        for zone, count_to_remove in broken_zones.items():
+            result += [x for x in brokers if x.zone == zone][:count_to_remove]
+        return result
+
+class FakeBroker(Broker):
+    def __init__(self):
+        super(FakeBroker, self).__init__(FAKE_BROKER_ID, FAKE_ZONE)
+        self.healthy = False
+
+    def add_partition(self, leader: bool, partition: Partition):
+        self._add_partition_int(partition.get_key(), leader)
+        return True
+
+
+def transfer_partition(partition, from_: Broker, target_brokers_sorted: List[Broker],
+                       zone_checker: ZoneChecker, polite=False):
+    for target in target_brokers_sorted:
+        if not zone_checker.is_allowed_move(from_, target, partition):
+            continue
+        if polite and target.is_overloaded(False, 1):
```

Comment:
should you actually pass the value from the caller function? it is not if it is leader or not at that point